### PR TITLE
Dynamic gpio specifcation from cmdline

### PR DESCRIPTION
**Pull request by penfold42**
This commit extends pi-blaster to let you specify the list of GPIO pins at run time from the command line using --gpio 4,7,25 .

If you omit this option, the default list of GPIOs is used so this should be backwards compatible.

I've also cleaned up some of the messages that are printed.

### CODE REVIEW

**Comment by penfold42** [comment]
I think this is ready to be merged now it's been tested

**Comment by penfold42** [comment]
also fixes #32

Any objections to merging ?

File: pi-blaster.c [hunk 1]
```diff
@@ -45,9 +45,11 @@ static char VERSION[] = "SNAPSHOT";
 #include <sys/mman.h>
 #include "mailbox.h"
 
-// Created new known_pins with raspberry pi list of pins
+#define MAX_CHANNELS	64
```
Comment (sarfata) [inline pi-blaster.c]: Why 64? We will never get to 64 because the GPIO register we write to is a uint32.

File: pi-blaster.c [hunk 2]
```diff
@@ -200,7 +223,7 @@ V revision (0-15)
 #define BOARD_REVISION_TYPE_CM (6 << 4)
 #define BOARD_REVISION_REV_MASK (0xF)
 
-#define LENGTH(x)  (sizeof(x) / sizeof(x[0]))
+//#define LENGTH(x)  (sizeof(x) / sizeof(x[0]))
```
Comment (sarfata) [inline pi-blaster.c]: Can you remove this?

File: pi-blaster.c [hunk 3]
```diff
@@ -240,8 +263,10 @@ static volatile uint32_t *gpio_reg;
 
 static int delay_hw = DELAY_VIA_PWM;
 static int invert_mode = 0;
+static int daemonize = 1;
 
-static float channel_pwm[NUM_CHANNELS];
+//static float channel_pwm[NUM_CHANNELS];
```
Comment (sarfata) [inline pi-blaster.c]: Please remove this line too.

File: pi-blaster.c [hunk 4]
```diff
@@ -57,17 +59,38 @@ static uint8_t known_pins[] = {
...
+		51,		// 47 to 53 are used by the SD card interface.
+		52,		// 47 to 53 are used by the SD card interface.
+		53,		// 47 to 53 are used by the SD card interface.
+};
 
 // Set num of possible PWM channels based on the known pins size.
-#define NUM_CHANNELS    (sizeof(known_pins)/sizeof(known_pins[0]))
+
+//#define NUM_CHANNELS    (sizeof(known_pins)/sizeof(known_pins[0]))
+uint8_t NUM_CHANNELS = (sizeof(known_pins)/sizeof(known_pins[0]));
```
Comment (sarfata) [inline pi-blaster.c]: This is not a constant anymore so you should rename it to be lower case: `num_channels`.

File: pi-blaster.c [hunk 5]
```diff
@@ -454,8 +492,10 @@ set_pin2gpio(int pin, float width){
 static void
 compact_pin2gpio(){
   int i, j = 0;
-  uint8_t tmp_pin2gpio[] = { 0,0,0,0,0,0,0,0 };
-  float tmp_channel_pwm[] = { 0,0,0,0,0,0,0,0 };
+//  uint8_t tmp_pin2gpio[] = { 0,0,0,0,0,0,0,0 };
+//  float tmp_channel_pwm[] = { 0,0,0,0,0,0,0,0 };
```
Comment (sarfata) [inline pi-blaster.c]: please remove commented code.

File: pi-blaster.c [hunk 6]
```diff
@@ -500,7 +540,8 @@ set_pin(int pin, float width)
   if (is_known_pin(pin)){
 	set_pin2gpio(pin, width);
   }else{
-	fprintf(stderr, "Not a known pin for pi-blaster\n");
+//	fprintf(stderr, "Not a configured pin for pi-blaster\n");
```
Comment (sarfata) [inline pi-blaster.c]: please remove.

File: pi-blaster.c [hunk 7]
```diff
@@ -878,14 +925,65 @@ parseargs(int argc, char **argv)
...
 				"-v (--version) - version information\n"
 				, argv[0]);
 			exit(-1);
 
+		case 'D':
+			daemonize = 0;
+			break;
+
+		case 'g':
+//			printf ("got gpio opt with arg %s\n", optarg);
```
Comment (sarfata) [inline pi-blaster.c]: please remove this line.

File: pi-blaster.c [hunk 8]
```diff
@@ -878,14 +925,65 @@ parseargs(int argc, char **argv)
...
+				i=0;
+				while( token != NULL ) 
+				{
+					if ((temp = atoi(token))) {
+//						printf( "Found gpio %d\n", temp);
+						if (i>=MAX_CHANNELS) {
+							printf( "ERROR maximum channels (%d) exceeded.\n", MAX_CHANNELS);
+							exit(-1);
+						}
+						if ((!is_banned_pin(temp)) && (temp>=0) && (temp<54) ){
```
Comment (sarfata) [inline pi-blaster.c]: What is 54 here? Can you make that a constant?

File: pi-blaster.c [hunk 9]
```diff
@@ -971,10 +1069,16 @@ main(int argc, char **argv)
...
+	if (daemonize) {
+		if (daemon(0,1) < 0) 
+			fatal("pi-blaster: Failed to daemonize process: %m\n");
+		else
+			printf("Daemonized, ");
+	}
+	printf("Reading %s.\n", DEVFILE);
 
-	printf("Initialisation finished, pi-blaster now running as daemon, waiting for input on %s\n", DEVFILE);
+//	printf("Initialisation finished, pi-blaster now running as daemon, waiting for input on %s\n", DEVFILE);
```
Comment (sarfata) [inline pi-blaster.c]: again, please remove dead code.

**Comment by sarfata** [comment]
Thanks for pushing on this. I have added a bunch of comments. As I have said before, I think we could do simpler but since I have not had time to do it, I will merge your PR when the comments are addressed.

Thanks!

**Comment by penfold42** [comment]
Thanks for the detailed feedback - I'd also have been happy with "clean up the old crap !"

So, if we're addressing the gpios with a uint32 I guess we'll never be able to hit the higher gpios that are present but not easily available on the board.
You want me to leave the defensive blacklist code in even tho it will never be used in the short term ?

**Comment by penfold42** [comment]
Done,

let me know if you'd like me to change anything else

**Comment by sarfata** [comment]
> So, if we're addressing the gpios with a uint32 I guess we'll never be able to hit the higher gpios that are present but not easily available on the board.

I just checked in the [datasheet](https://www.raspberrypi.org/wp-content/uploads/2012/02/BCM2835-ARM-Peripherals.pdf), page 90 and 95, you can see that there are two GPIO set registers. One for GPIO from 0 to 31 and one for GPIO 32 to 53. (Same thing for function selection and clear)

Right now [we only support writing to the first register](https://github.com/penfold42/pi-blaster/blob/dynamic-gpio/pi-blaster.c#L310). I do not think setting a GPIO above 32 works. But if we just changed our code a little bit, we could support it. I will create a new ticket for this (#66).

> You want me to leave the defensive blacklist code in even tho it will never be used in the short term ?

I am not sure what you are referring to. Can you explain?

I will merge this now. Thank you very much for your help!

**Comment by penfold42** [comment]
There are a number of gpios > 31 that are used be HDMI and the sdcard interface.

Because we're hitting the hardware directly I'm not sure if this would corrupt reads and writes to the SX card so i created a list that I refuse to let the user use.

http://elinux.org/RPi_Low-level_peripherals

**Comment by sarfata** [comment]
Oh yes I get it. You were right to add that banned_list and yes we should keep it. People can remove the banned pins if they really know what they are doing.

**Pull request merged**